counsel.el

```diff
@@ -5394,6 +5432,46 @@ specified by the `blddir' property."
             :action #'counsel-compile--action
             :caller 'counsel-compile))
 
+
+;(defun ivy--format-function-generic (selected-fn other-fn cands separator)
+(defun counsel-compile-env--format-hint (cands)
+  "Return a formatter for compile-env CANDS."
+  (ivy--format-function-generic
+   (lambda (selected)
+     (format "%s %s"
+             (if (member selected counsel-compile-env)
+                 (propertize "remove" 'face 'font-lock-warning-face)
+               (propertize "add" 'face
+                           'font-lock-variable-name-face))
+             selected))
+   (lambda (other)
+     other)
+   cands
+   "\n"))
+
+(defun counsel-compile-env--update (var)
+  "Update `counsel-compile-env' either adding or removing VAR."
+  (if (member var counsel-compile-env)
+      (setq counsel-compile-env (delete var counsel-compile-env))
+    (if (string-match-p counsel-compile-env-pattern var)
+        (add-to-list 'counsel-compile-env var)
+      (message "Did not add badly formed variable '%s' to environment" var))))
+
+;;;###autoload
+(defun counsel-compile-env ()
+  "Update `counsel-compile-env' interactively."
+  (interactive)
+  (let ((ivy-format-function #'counsel-compile-env--format-hint))
+    (ivy-read "Compile environment variable: " (delete-dups
+                                                (append
+                                                 counsel-compile-env counsel-compile-env-history))
```

Comment:
Nitpick: This line is getting pretty long; please move the second argument to `ivy-read` to a separate line.